### DISCUSSION TITLE
Coloring of single objects in Tile3DLayer

### DISCUSSION THREAD

**Discussion by baal-lgln** [visgl/deck.gl#6151] [Q&A]
Currently I'm adding I3S data (buildings) with a Tile3DLayer to my map. Picking and highlighting works. After selecting/picking a building I can also display the attributes of the correspinding building. Now I'm wondering if it is already possible to colorize the selected building (or even all buildings without selecting) according to a given attribute (e.g. measuredHeight)? 

```javascript
const buildingsLayer = new Tile3DLayer({
    id: "3d-buildings",
    data: TILESET_URL,
    loader: I3SLoader,
    pickable: true,
    highlightedObjectIndex: selectedBuilding,
    autoHighlight: true
});

const [selectedBuilding, setSelectedBuilding] = useState(-1);

const deckRef = useRef(null);

const onClick = useCallback(event => {
    const pickInfo = deckRef.current.pickObject({
        unproject3D: true, // needed to get the z value in pickInfo.coordinate
        x: event.clientX,
        y: event.clientY,
        radius: 1
    });
    if (pickInfo) {
        if (pickInfo.layer.id === "3d-buildings") {
            const selectFeatureAttributes = async () => {
                const buildingFeatureAttributes = await loadFeatureAttributes(pickInfo.object, pickInfo.index);
                console.log(buildingFeatureAttributes);
                setSelectedBuilding(pickInfo.index);
            }
            selectFeatureAttributes();
        }
    }
}, []);
```

Has this been done or are there any examples? The highlighting function seems to do something similiar but I don't see any way to do this myself in the docs. Is maybe using updateTriggers the right way to go here? But what accessors could I use?

**Comment by mcost45** [comment]
Some layers offer some basic colour altering properties (like bitmap layer), but I don't know if any exist for your layers. You will likely have to extend the [specific Tile3DLayer sub layer your format is using](https://observablehq.com/@pessimistress/deck-gl-shader-injection) to inject some shader/glsl code, and also extend the Tile3DLayer so it uses the new sub layer class (instead of the original, for compositing).

Customising layer shaders gives you a great amount of control over rendered colours. You can do this with [Deck's shader injection](https://deck.gl/docs/developer-guide/custom-layers/writing-shaders) - [check out this example](https://observablehq.com/@pessimistress/deck-gl-shader-injection). You will probably want to use the vs:DeckGL_FILTER_COLOR or fs:DeckGL_FILTER_COLOR hooks, depending on what you are trying to do. This way, you can alter colours depending on some attributes passed into the shaders.

It is a bit of a learning curve if you're not used GLSL/shaders/attributes/uniforms etc. It can be helpful looking at the vertex/fragment shader source code for the layer you are injecting, for a better understanding of the uniforms/attributes already in use.
